File: 11-container-with-most-water/container-with-most-water.java
/*class Solution{

    public static int maxArea(int[] height) {

        int maxWater = 0;

        for (int i = 0; i < height.length; i++) {

            for (int j = i + 1; j < height.length; j++) {

                int width = j - i;

                int minHeight = Math.min(height[i], height[j]);

                int currentWater = width * minHeight;

                if (currentWater > maxWater) {
                    maxWater = currentWater;//brute force approach
                }
            }
        }

        return maxWater;
    }

    public static void main(String[] args) {

        int[] height = {1, 8, 6, 2, 5, 4, 8, 3, 7};

        System.out.println(maxArea(height));
    }
}*/


class Solution {

    public int maxArea(int[] height) {

        int left = 0;
        int right = height.length - 1;

        int maxWater = 0;

        while (left < right) {

            int width = right - left;

            int minHeight = Math.min(height[left], height[right]);//optimized approach

            int currentWater = width * minHeight;

            maxWater = Math.max(maxWater, currentWater);//time complexity:O(n)

            // Move smaller pointer
            if (height[left] < height[right]) {
                left++;
            } else {
                right--;
            }
        }

        return maxWater;
    }
}

File: 11-container-with-most-water/container-with-most-water.py
class Solution:
    # height is the input array
    def maxArea(self, height):

        # Left pointer starts from the first index
        left = 0

        # Right pointer starts from the last index
        right = len(height) - 1

        # Stores the maximum area found so far
        max_water = 0

        # Continue until both pointers meet
        while left < right:

            # Width is the distance between the indices
            width = right - left

            # The shorter wall limits the water level
            current_height = min(height[left], height[right])

            # Calculate the current area
            area = width * current_height

            # Update the maximum area if the current area is larger
            max_water = max(max_water, area)

            # Move the pointer pointing to the shorter wall
            if height[left] < height[right]:
                left += 1
            else:
                right -= 1

        # Return the maximum water stored
        return max_water